src/lambda_vm/interpreter.ml

```diff
@@ -0,0 +1,192 @@
+
+  val script_result : (value * (nil * nil)) t
+  val parse : value -> 'a t -> 'a
+end = struct
+  type nil = unit
+
+  type 'a t =
+    | Any : value t
+    | Pair : 'left t * 'right t -> ('left * 'right) t
+    | Nil : nil t
+
+  let any = Any
+  let pair left right = Pair (left, right)
+  let nil = Nil
+
+  let operations = nil
+  let script_result = pair any (pair operations nil)
+
+  let rec parse : type a. value -> a t -> a =
+   fun value t ->
+    match (t, value) with
+    | Any, value -> value
+    | Pair (left_t, right_t), V_pair { left = left_value; right = right_value }
+      ->
+      let left = parse left_value left_t in
+      let right = parse right_value right_t in
+      (left, right)
+    | Pair _, (V_int64 _ | V_closure _ | V_primitive _) ->
+      raise Value_is_not_pair
+    | Nil, V_int64 0L -> ()
+    | Nil, (V_int64 _ | V_pair _ | V_closure _ | V_primitive _) ->
+      raise Value_is_not_zero
+end
+
+module Env = Map_with_cardinality.Make (Ident)
+
+let burn_gas gas env code =
+  (match code with
+  | E_var _
+  | E_app _ ->
+    let cardinality = Env.cardinal env in
+    Gas.burn_log2 gas ~cardinality
+  | E_lam _
+  | E_const _
+  | E_prim _
+  | E_if _
+  | E_pair _
+  | E_fst _
+  | E_snd _ ->
+    Gas.burn_constant gas);
+  check_gas gas
+
+let eval_prim prim ~arg ~args =
+  let op1 f =
+    let f value =
+      match value with
+      | V_int64 value -> V_int64 (f value)
+      | V_pair _
+      | V_closure _
+      | V_primitive _ ->
+        raise Value_is_not_int64 in
+    match args with
+    | [] -> f arg
+    | _ -> raise Over_applied_primitives in
+
+  let op2 f =
+    (* error only happens after both are applied *)
+    let f left right =
+      match (left, right) with
+      | V_int64 left, V_int64 right -> V_int64 (f left right)
+      | ( (V_pair _ | V_int64 _ | V_closure _ | V_primitive _),
+          (V_pair _ | V_int64 _ | V_closure _ | V_primitive _) ) ->
+        raise Value_is_not_int64 in
+    match args with
+    | [] -> V_primitive { args = [arg]; prim }
+    | [left] -> f left arg
+    | _ -> raise Over_applied_primitives in
+
+  let op2_shift f =
+    let f left right = f left (Int64.to_int right) in
+    op2 f in
+  match prim with
+  | P_neg -> op1 Int64.neg
+  | P_add -> op2 Int64.add
+  | P_sub -> op2 Int64.sub
+  | P_mul -> op2 Int64.mul
+  | P_div -> op2 Int64.div
+  | P_rem -> op2 Int64.rem
+  | P_and -> op2 Int64.logand
+  | P_or -> op2 Int64.logor
+  | P_xor -> op2 Int64.logxor
+  | P_lsl -> op2_shift Int64.shift_left
+  | P_lsr -> op2_shift Int64.shift_right_logical
+  | P_asr -> op2_shift Int64.shift_right
+
+(* TODO: gas must be ref *)
+let rec eval ~stack gas env code =
+  let eval_call env code = eval ~stack:(stack - 1) gas env code in
+  let eval_jump env code = eval ~stack gas env code in
+
+  check_stack ~stack;
+  burn_gas gas env code;
+
+  match code with
+  | E_var var -> (
+    (* TODO: gas cost for this *)
+    (* TODO: shuold this be a hashtbl? Or an array? *)
```

Comment:
to arrayify this you will need to convert to de bruijn indices, a hash table sounds bad bc functional insertions are slow